src/azul/deployment.py

```diff
@@ -277,5 +285,149 @@ def _sanitize_tf(tf_config: JSON) -> JSON:
     return {k: v for k, v in tf_config.items() if v}
 
 
+def _flatten_nested_block(block: Union[Dict, List[Dict]]) -> Iterator[Tuple[str, Any]]:
+    """
+    Some Terraform blocks can be directly nested or encoded as a list. For
+    example, these are equivalent:
+
+        {"resource": {"resource_type": {"resource_id": {"foo": ...}}}}
+        {"resource": [{"resource_type": {"resource_id": {"foo": ...}}}]}
+
+    Each child dict can hold multiple keys that need to be regarded separately.
+    For example, these are equivalent:
+
+        {"resource": {"type": {"id": {"foo": ...}, "id2": {"bar": ...}}}}
+        {"resource": {"type": [{"id": {"foo": ...}}, {"id2": {"bar": ...}}]}}
+
+    This helper function normalizes input to prefer the second form of both
+    cases to make parsing Terraform easier.
+
+    >>> list(_flatten_nested_block({"foo": "bar"}))
+    [('foo', 'bar')]
+
+    >>> list(_flatten_nested_block([{"foo": "bar"}]))
+    [('foo', 'bar')]
+
+    >>> list(_flatten_nested_block({"foo": "bar", "baz": "qux"}))
+    [('foo', 'bar'), ('baz', 'qux')]
+
+    >>> list(_flatten_nested_block([{"foo": "bar"}, {"baz": "qux"}]))
+    [('foo', 'bar'), ('baz', 'qux')]
+
+    >>> list(_flatten_nested_block([{"foo": "bar", "baz": "qux"}]))
+    [('foo', 'bar'), ('baz', 'qux')]
+    """
+    if isinstance(block, dict):
+        return itertools.chain(block.items())
+    else:
+        return itertools.chain.from_iterable(attr.items() for attr in block)
+
+
+def _populate_tags(tf_config: JSON) -> JSON:
+    """
+    Adds tags to all resources and changes the `name` tag to `Name` for tagged
+    AWS resources.
+
+    >>> _populate_tags({"resource": [{"aws_vpc": {"res_name": {"tags": {"name": "foo"}}}}]})
+    {'resource': [{'aws_vpc': {'res_name': {'tags': {'Name': 'foo'}}}}]}
+
+    >>> _populate_tags({"resource": [{"google_service_account": {"res_name": {"tags": {"name": "foo"}}}}]})
+    {'resource': [{'google_service_account': {'res_name': {'tags': {'name': 'foo'}}}}]}
+    """
+    if 'resource' not in tf_config:
+        return tf_config
+    resources = []
+    for resource_type, resource in _flatten_nested_block(tf_config['resource']):
+        for resource_name, properties in _flatten_nested_block(resource):
+            properties = add_tags(resource_type, properties)
+            if 'tags' in properties and 'name' in properties['tags'] and resource_type.startswith('aws_'):
+                properties['tags']['Name'] = properties['tags'].pop('name')
+            resources.append({resource_type: {resource_name: properties}})
+    return {**tf_config, 'resource': resources}
+
+
 def emit_tf(tf_config: Optional[JSON]):
-    return emit(tf_config) if tf_config is None else emit(_sanitize_tf(tf_config))
+    return emit(tf_config) if tf_config is None else emit(_sanitize_tf(_populate_tags(tf_config)))
+
+
+def tags(resource_name: str, **kwargs) -> Dict[str, str]:
```

Comment:
It seems like a lot of the arguments you pass in for `resource_name` aren't parsed correctly by `config.unqualified_resource_name_or_none()`. `azul-gitlab` failed for me, but you should test the others too.